### ISSUE TITLE
Avoid blocking the map in a distributed (re)loading [HZ-1177]

### ISSUE THREAD

**Issue by jmatas** [hazelcast/hazelcast#9467]
It would be nice to have the possibility to perform the map loading in two modes:
- blocking the client requests (current behaviour)
- non-blocking the client requests (this would be the new mode).

**Use case***: 
Currently, I have an IMap backed with a MapStore. This IMap is loaded on startup and have millions of entries. When the map is fully loaded, the clients retrieves entries using Imap.values() with predicates. This loading work fine for me on startup.

Now, I have to reload peridiocally all the data (e.g., every 2 hours). I'm using IMap.loadAll(true) to force the reload. The MapStore (loadAllKeys() + loadAll(keys)) takes about 4 minutes to finish the complete load. It seems that in these 4 minutes, all the clients that send queries to the imap via predicates are blocked until the reloading is finished.

In my case, I can't block the cluster 4 minutes (neither a second in fact).
- is really the Imap.loadAll blocking the predicates queries?
- is possible configure the imap as non-blocking or exists a non-blocking alternative to IMap.loadAll().

I'm trying it with HC 3.7.2, with 2 nodes and java 8.

*from: http://stackoverflow.com/questions/41153960/is-hazelcast-distributed-loading-blocking

**Comment by jmatas** [comment]
And it would be even nicer if the map could be configured as auto-reloaded. For example, can configure a period of 2 hours and hazelcast manage automatically the reload every two hours, without an external scheduler that calls to IMap.loadAll(true).

**Comment by pablo-u** [comment]
+1. it'd be great to have this feature available. I'm experiencing the same issue and can't afford to delay any of my requests during reload.

**Comment by krasnyt** [comment]
+1

**Comment by SearchEngine27** [comment]
What I've observed is that when you have two completely distinct IMaps (map A and B), with two completely distinct MapLoaders (MapLoaders A->X and B->Y), whereby both are distributed data structures within the same cluster, that a get operation on map A is blocked by a MapLoader operation on loader Y. Should this be a separate bug?

**Comment by bastrich** [comment]
+ 1

**Comment by bastrich** [comment]
`+ 1

**Comment by Holmistr** [comment]
@bastrich @krasnyt @pabloubal @eric-eldar @jmatas I know it's been a long time but we're finally implementing the offloading of MapLoader/Store operations out of the partition thread in 5.2. Stay tuned!

@ahmetmircik Could you please take a look at this comment https://github.com/hazelcast/hazelcast/issues/9467#issuecomment-505909477 ?

**Comment by ahmetmircik** [comment]
@Holmistr they are served by same hazelcast threads, even you have distinct IMaps, they share these threads. So one slow loading can block another map's requests.

**Comment by Holmistr** [comment]
I see. Therefore, offloading the MapLoader operations will solve this as well as the other map's operations will also be able to proceed. Cool!

**Comment by AyberkSorgun** [comment]
Done via https://github.com/hazelcast/hazelcast/pull/21651
